### DISCUSSION TITLE
Mix npm public repositories with private gitlab npm

### DISCUSSION THREAD

**Discussion by hieroshima** [renovatebot/renovate#8413] [Request Help] [answered]
Hi,

I would like to ask if possible is to mix private npm on gitlab with public npm registry? Currently, I have that kind of configuration:

```
...
"ignoreNpmrcFile": true,
"hostRules": [
    {
      "hostName": "gitlab.com",
      "ignoreNpmrcFile": false
    }
  ],
...
```
and I'm using self-hosted Renovate installation. Inside the main dir where config.js is, I have .npmrc file with auth to the gitlab.com npm repo. Once I'm running a bot, I'm getting error:

WARN: artifactErrors (repository=<repo-where-bot-is-running>, branch=dependencies/update-dependencies-<name-of-lib-on-gitlab-npm>)
       "artifactErrors": [
         {
           "lockFile": "src/<repo-name>/package-lock.json",
           "stderr": "npm ERR! code E404\nnpm ERR! 404 Not Found - GET **https://registry.npmjs.org/[name-of-the-private-repository-on-gitlab] - Not found\nnpm ERR! 404** \nnpm ERR! 404  **[private-repository-on-gitlab]@^0.31.0' is not in the npm registry.\nnpm ERR! 404** "

I think that .npmrc is overwritten npm registry, but that is why I used ignoreNpmrcFile for all and used hostRules for gitlab.

Can someone advice how it should be done to work in that way? :)

Thanks in advance

**Comment by viceice** [comment, marked as answer]
`ignoreNpmrcFile` can't be used in host rules, it is a top level config option only

**Reply by viceice** [reply]
Can you please describe what are you want to have updated? a minimal sample repo is helpful too.

**Reply by hieroshima** [reply]
@viceice thanks for answer. So, let me show you my package.json. In the dependency you can see that I have one library from gitlab npm which is a private repository available for me and also I have few other libraries from public npm registry.

```
{
  "name": "my-package",
  "version": "1.0.0",
  "description": "",
  "main": "app.ts",
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "reflect-metadata": "^0.1.13",
    "typedi": "^0.8.0"
  },
  "dependencies": {
    "@<private-gitlab-repo>/shared-library": "^0.18.0",
  }
}
```
When I run Bot, he will create a 3 MR on gitlab, but the MR where shared-library is for an update is not mergeable, because of the error. The rest two, which are from npm registry, are mergeable. It look like it tries to get library from public npm rather than gitlab.

![image](https://user-images.githubusercontent.com/6417562/105717274-37853100-5f20-11eb-89e6-f094a8960249.png)

**Reply by hieroshima** [reply]
When I use npmrc inside renovate.json config file inside the repository, it does not create a MR with that private library.

**Reply by viceice** [reply]
You need to add a host rule with the token, so renovate can pass it to the lockfile processing (it will add it to the repo `.npmrc` temporary)

renovate config:
```js
module.exports = {
  hostRules: [
   {
      hostType: 'npm',
      baseUrl: 'https://npm.xyz.de/',
      token: process.env.NPM_XYZ_TOKEN,
    }
  ]
}
```
and in you repo `.npmrc` you just need
```ini
@private-scope:registry="https://npm.xyz.de/"
```

**Reply by hieroshima** [reply]
Thanks :) - is there a possibility to do something similar but with packageName? Because I would like to avoid updating package.json by adding registry, I would prefer to use a package name which is usually mapped in .npmrc file.

**Reply by viceice** [reply]
Updated my comment, the registry mapping should already be in your repo `.npmrc` 🙃 

So you only need to add the host rule and no other npm related settings in bot config

**Reply by rarkins** [reply]
Isn't `.npmrc` fully supportive of package scopes? e.g. default to registry.npmjs.org but redirect `@foobar` to your private registry? Then as @viceice mentions, add a `hostRules` entry which gives the token for your private registry, and it will be appended to the `.npmrc` before we run `npm install`.

**Reply by hieroshima** [reply]
It does not work for me. When I put host rules into configuration it not see my repository. In console I see that it recognise it as an gitlabci not an npm package:
```
managers": {
           "gitlabci": {"fileCount": 1, "depCount": 1},
           "gitlabci-include": {"fileCount": 1, "depCount": 2},
           "npm": {"fileCount": 2, "depCount": 19}
         }
```
Maybe this is a problem? Now it does not even create an MR

**Reply by viceice** [reply]
Please share your bot config and the `renovate.json` of the repo.

**Reply by hieroshima** [reply]
Bot config on the machine (config.js):

```javascript
module.exports = {
  endpoint: 'https://gitlab.com/api/v4/',
  token: '<access-token>',
  platform: 'gitlab',
  hostRules: [
   {
      hostType: 'npm',
      baseUrl: 'https://gitlab.com',
      token: '<npm-token>'
    }
  ],
  optimizeForDisabled: true,
  trustLevel: 'high',
  onboardingConfig: {
    extends: ['config:base'],
  },
  repositories: [
    <repo-lists>
  ],
};
```
Repository config (renovate.json):

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":ignoreModulesAndTests",
    ":semanticCommits",
    ":disableRateLimiting",
    ":automergeDisabled",
    ":maintainLockFilesDisabled",
    ":ignoreUnstable",
    ":semanticCommitTypeAll(BOT)",
    ":preserveSemverRanges"
  ],
  "separateMultipleMajor": false,
  "separateMinorPatch": false,
  "packageRules": [
    {
      "depTypeList": ["devDependencies"],
      "branchPrefix": "devDependencies/update-{{depName}}-",
      "groupName": "devDependencies",
      "groupSlug": "dev-dependencies",
      "labels": ["devDependency"]
    },
    {
      "depTypeList": ["dependencies"],
      "branchPrefix": "dependencies/update-{{depName}}-",
      "groupName": "dependencies",
      "groupSlug": "dependencies",
      "labels": ["dependency"]
    },
    {
      "updateTypes": ["major"],
      "addLabels": ["major"]
    },
    {
      "updateTypes": ["minor"],
      "addLabels": ["minor"]
    },
    {
      "updateTypes": ["patch"],
      "addLabels": ["patch"]
    }
  ],
  "commitBodyTable": true,
  "prHeader": "{{#if isMajor}}# {{depName}} v{{{newMajor}}} &ndash;&ndash; :bangbang: MAJOR UPDATE! :bangbang:{{/if}}",
  "commitMessageSuffix": "{{#if isMajor}}[MAJOR]{{/if}}"
}

```

**Reply by viceice** [reply]
Can you please share the `DEBUG` logs. There should be a line `DEBUG: packageFiles with updates` and a big json. There you should find all found dependencies.

You should also find messages if some deps version lookup failed.

**Reply by rarkins** [reply]
And what about any .npmrc in the repo?

**Reply by hieroshima** [reply]
.npmrc in the repo

```
@<private-repo>:registry=https://gitlab.com/api/v4/packages/npm/
```

I can't see debug logs I see only INFO and Warnings and warnings is the same which I had at the beginning - it tries get my private repo from an npm

**Reply by viceice** [reply]
@hieroshima set your env to `LOG_LEVEL=debug` before running renovate to get debug logs

**Reply by rarkins** [reply]
> .npmrc in the repo
> 
> ```
> @<private-repo>:registry=https://gitlab.com/api/v4/packages/npm/
> ```
> 
> I can't see debug logs I see only INFO and Warnings and warnings is the same which I had at the beginning - it tries get my private repo from an npm

Please make sure it is scope, not repo, e.g.:

```
@myscope:registry=https://mycustomregistry.example.org
```

**Reply by hieroshima** [reply]
@rarkins  Sorry, I used wrong word - it is the scope - not the repository :)

@viceice debug:

```javascript
{
                 "depType": "dependencies",
                 "depName": "@<private-scope>/shared-library",
                 "currentValue": "^0.18.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": "0.18.0",
                 "depIndex": 0,
                 "updates": [
                   {
                     "fromVersion": "0.18.0",
                     "toVersion": "0.31.0",
                     "newValue": "^0.31.0",
                     "newMajor": 0,
                     "newMinor": 31,
                     "updateType": "minor",
                     "isSingleVersion": false,
                     "isRange": true
                   }
                 ],
                 "warnings": [],
                 "fixedVersion": "0.18.0"
               },
```
Error:

```javascript
DEBUG: Executing command
"command": "npm install --package-lock-only --ignore-scripts --no-audit"
DEBUG: lock file error 
"err": {
"killed": false,
"code": 1,
"signal": null,
"cmd": "npm install --package-lock-only --ignore-scripts --no-audit",
"stdout": "",
"stderr": "npm ERR! code E404\nnpm ERR! 404 Not Found - GET https://**redacted**@<private-scope>%2fshared-library\nnpm ERR! 404 \nnpm ERR! 404  '@<private-scope>/shared-library@^0.31.0' is not inthe npm registry.\nnpm ERR! 404 You should bug the author to publish it (or use the name yourself!)\nnpm ERR! 404 It was specified as a dependency of 'team-assets'\nnpm ERR! 404 \nnpm ERR! 404 Note that you canalso install from a\nnpm ERR! 404 tarball, folder, http url, or git url.\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate/cache/others/npm/_logs/2021-01-26T09_59_37_647Z-debug.log\n",
"message": "Command failed: npm install --package-lock-only --ignore-scripts --no-audit\nnpm ERR! code E404\nnpm ERR! 404 Not Found - GET https://**redacted**@<private-scope>%2fshared-library\nnpm ERR! 404 \nnpm ERR! 404  '@<private-scope>/shared-library@^0.31.0' is not in the npm registry.
```

When I open this 404 link - it is to gitlab npm with my library - I can see whole library with all versions

**Reply by hieroshima** [reply]
I don't have an idea how it is possible that he knows about to what version this lib should be upgraded, but at the same time it returns 404 error that lib does not exist

**Reply by rarkins** [reply]
This means that *Renovate* knows how to look up the private package but now `npm` does not. i.e. the `.npmrc` which `npm` is accessing does not contain the right registry and/or token.

**Reply by rarkins** [reply]
Please make sure that the `baseUrl` in your `hostRules` is `https://gitlab.com/api/v4/packages/npm/` and not the shortened `https://gitlab.com`

**Reply by hieroshima** [reply]
🤦  How I could miss this... Thank you so much! It works now perfectly :)
